rust/cardano-chain-follower/src/follow.rs

```diff
@@ -85,53 +85,54 @@ impl ChainFollower {
-                mithril_tip = ?current_mithril_tip,
-                "Setting Initial Mithril Tip"
-            );
-            self.mithril_tip = Some(current_mithril_tip.clone());
-            &current_mithril_tip
-        };
-
-        // Return an ImmutableBlockRollForward event as soon as we can after one occurs.
-        // This is not an advancement in the followers sequential block iterating state.
-        // BUT it is a necessary status to return to a follower, so it can properly handle
-        // when immutable state advances (if it so requires)
-        if current_mithril_tip != *previous_mithril_tip {
-            debug!(
-                new_tip = ?self.mithril_tip,
-                current_tip = ?current_mithril_tip,
-                "Mithril Tip has changed"
-            );
-            // We have a new mithril tip so report Mithril Tip Roll Forward
-            if let Some(block) = self.snapshot.read_block_at(&current_mithril_tip).await {
-                // Update the snapshot in the follower state to the new snapshot.
-                let update = ChainUpdate::new(
-                    chain_update::Kind::ImmutableBlockRollForward,
-                    false, // Tip is Live chain tip, not Mithril Tip, and this is Mithril Tip.
-                    block,
+        loop {
+            let current_mithril_tip = latest_mithril_snapshot_id(self.chain).tip();
+
+            // Get previous mithril tip, or set it and return current if a previous does not exist.
+            let previous_mithril_tip = self.mithril_tip.get_or_insert_with(|| {
+                debug!(
+                    mithril_tip = ?current_mithril_tip,
+                    "Setting Initial Mithril Tip"
+                );
+                current_mithril_tip.clone()
+            });
+
+            // Return an ImmutableBlockRollForward event as soon as we can after one occurs.
+            // This is not an advancement in the followers sequential block iterating state.
+            // BUT it is a necessary status to return to a follower, so it can properly handle
+            // when immutable state advances (if it so requires)
+            if current_mithril_tip != *previous_mithril_tip {
+                debug!(
+                    new_tip = ?self.mithril_tip,
+                    current_tip = ?current_mithril_tip,
+                    "Mithril Tip has changed"
+                );
+                // We have a new mithril tip so report Mithril Tip Roll Forward
+                if let Some(block) = self.snapshot.read_block_at(&current_mithril_tip).await {
+                    // Update the snapshot in the follower state to the new snapshot.
+                    let update = ChainUpdate::new(
+                        chain_update::Kind::ImmutableBlockRollForward,
+                        false, // Tip is Live chain tip, not Mithril Tip, and this is Mithril Tip.
+                        block,
+                    );
+                    return Some(update);
```

Comment:
```suggestion
                    self.mithril_follower = None; // Next iteration needs to get a new mithril follower, drop the current one.
                    return Some(update);
```

Reply:
Now that the failsafe logic has been shown to work, this is the better correction as it should eliminate more than 99% of all mithril iterator errors when the mithril chain rolls forward.